src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpRequestCallback.cs

```diff
@@ -0,0 +1,360 @@
+            }
+        }
+
+        private static void OnRequestSendRequestComplete(WinHttpRequestState state)
+        {
+            Debug.Assert(state != null, "OnRequestSendRequestComplete: state is null");
+            Debug.Assert(state.TcsSendRequest != null, "OnRequestSendRequestComplete: TcsSendRequest is null");
+            Debug.Assert(!state.TcsSendRequest.Task.IsCompleted, "OnRequestSendRequestComplete: TcsSendRequest.Task is completed");
+            
+            state.TcsSendRequest.TrySetResult(true);
+        }
+
+        private static void OnRequestReadComplete(WinHttpRequestState state, uint bytesRead)
+        {
+            Debug.Assert(state != null, "OnRequestReadComplete: state is null");
+            Debug.Assert(state.TcsReadFromResponseStream != null, "TcsReadFromResponseStream is null");
+            Debug.Assert(!state.TcsReadFromResponseStream.Task.IsCompleted, "TcsReadFromResponseStream.Task is completed");
+            
+            state.TcsReadFromResponseStream.TrySetResult((int)bytesRead);
+        }
+
+        private static void OnRequestWriteComplete(WinHttpRequestState state)
+        {
+            Debug.Assert(state != null, "OnRequestWriteComplete: state is null");
+            Debug.Assert(state.TcsInternalWriteDataToRequestStream != null, "TcsInternalWriteDataToRequestStream is null");
+            Debug.Assert(!state.TcsInternalWriteDataToRequestStream.Task.IsCompleted, "TcsInternalWriteDataToRequestStream.Task is completed");
+            
+            state.TcsInternalWriteDataToRequestStream.TrySetResult(true);
+        }
+
+        private static void OnRequestReceiveResponseHeadersComplete(WinHttpRequestState state)
+        {
+            Debug.Assert(state != null, "OnRequestReceiveResponseHeadersComplete: state is null");
+            Debug.Assert(state.TcsReceiveResponseHeaders != null, "TcsReceiveResponseHeaders is null");
+            Debug.Assert(!state.TcsReceiveResponseHeaders.Task.IsCompleted, "TcsReceiveResponseHeaders.Task is completed");
+
+            state.TcsReceiveResponseHeaders.TrySetResult(true);
+        }
+
+        private static void OnRequestRedirect(WinHttpRequestState state, Uri redirectUri)
+        {
+            const string EmptyCookieHeader = "Cookie:";
+
+            Debug.Assert(state != null, "OnRequestRedirect: state is null");
+            Debug.Assert(redirectUri != null, "OnRequestRedirect: redirectUri is null");
+            Debug.Assert(state.TcsReceiveResponseHeaders != null, "TcsReceiveResponseHeaders is null");
+            Debug.Assert(!state.TcsReceiveResponseHeaders.Task.IsCompleted, "TcsReceiveResponseHeaders.Task is completed");
+
+            // If we're manually handling cookies, we need to reset them based on the new URI.
+            if (state.Handler.CookieUsePolicy == CookieUsePolicy.UseSpecifiedCookieContainer)
+            {
+                // Clear cookies.
+                if (!Interop.WinHttp.WinHttpAddRequestHeaders(
```

Comment:
Calling WinHTTP here could cause internal state issues. This probably requires acquiring the `state.Lock`.